src/backend/codegen/include/codegen/base_codegen.h

```diff
@@ -0,0 +1,158 @@
+   * @brief Destroys the code generator and reverts back to using regular
+   * 		version of the generated function.
+   **/
+  virtual ~BaseCodeGen() {
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+  }
+
+  bool GenerateCode(gpcodegen::CodeGenUtils* codegen_utils) final {
+    is_generated_ = DoCodeGeneration(codegen_utils);
+    return is_generated_;
+  }
+
+  bool SetToRegular() final {
+    assert(nullptr != regular_func_ptr_);
+    SetToRegular(regular_func_ptr_, ptr_to_chosen_func_ptr_);
+    return true;
+  }
+
+  bool SetToGenerated(gpcodegen::CodeGenUtils* codegen_utils) final {
+    if (false == IsGenerated()) {
+      assert(*ptr_to_chosen_func_ptr_ == regular_func_ptr_);
+      return false;
+    }
+
+    auto compiled_func_ptr = codegen_utils->GetFunctionPointerTypeDef<
+        FuncPtrType>(GetUniqueFuncName());
+
+    if (nullptr != compiled_func_ptr) {
+      *ptr_to_chosen_func_ptr_ = compiled_func_ptr;
+      return true;
+    }
+    return false;
+  }
+
+  void Reset() final {
+    SetToRegular();
+  }
+
+  const std::string& GetOrigFuncName() const final {
+    return orig_func_name_;
+  }
+
+  const std::string& GetUniqueFuncName() const final {
+    return unique_func_name_;
+  }
+
+  bool IsGenerated() const final {
+    return is_generated_;
+  }
+
+  /**
+   * @return Regular version of the corresponding generated function.
+   *
+   **/
+  FuncPtrType GetRegularFuncPointer() {
+    return regular_func_ptr_;
+  }
+
+  /**
+   * @brief Sets up the caller to use the corresponding regular version of the
+   *        generated function.
+   *
+   * @param regular_func_ptr       Regular version of the generated function.
+   * @param ptr_to_chosen_func_ptr Reference to caller.
+   *
+   * @return true on setting to regular version.
+   **/
+  static bool SetToRegular(FuncPtrType regular_func_ptr,
+                           FuncPtrType* ptr_to_chosen_func_ptr) {
+    *ptr_to_chosen_func_ptr = regular_func_ptr;
+    return true;
+  }
+
+ protected:
+  /**
+   * @brief Constructor
+   *
+   * @param orig_func_name         Original function name.
+   * @param regular_func_ptr       Regular version of the generated function.
+   * @param ptr_to_chosen_func_ptr Reference to the function pointer that the caller will call.
+   *
+   * @note 	The ptr_to_chosen_func_ptr can refer to either the generated function or the
+   * 			corresponding regular version.
+   *
+   **/
+  explicit BaseCodeGen(const std::string& orig_func_name,
+                       FuncPtrType regular_func_ptr,
+                       FuncPtrType* ptr_to_chosen_func_ptr)
+  : orig_func_name_(orig_func_name),
+    unique_func_name_(CodeGenInterface::GenerateUniqueName(orig_func_name)),
+    regular_func_ptr_(regular_func_ptr),
+    ptr_to_chosen_func_ptr_(ptr_to_chosen_func_ptr),
+    is_generated_(false) {
+    // Initialize the caller to use regular version of generated function.
+    SetToRegular(regular_func_ptr, ptr_to_chosen_func_ptr);
```

Comment:
As I mentioned above, let's introduce a default manager, and then we can make all the SetToRegular and SetToGenerated protected (or even private if we can swap the pointers in the base code gen)